### PR TITLE
RUBY-2371: Use monotonic clock for measuring time

### PR DESCRIPTION
**Pull request by comandeo**
This PR introduces usage of monotonic clock instead of wall clock for measuring time. This ensures correct calculation of elapsed times.

See https://blog.dnsimple.com/2018/03/elapsed-time-with-ruby-the-right-way/ for details

### CODE REVIEW

File: lib/mongo/cluster.rb [hunk 1]
```diff
@@ -214,6 +214,7 @@ def initialize(seeds, monitoring, options = Options::Redacted.new)
 
       # Need to record start time prior to starting monitoring
       start_time = Time.now
```
Comment (p-mongo) [inline lib/mongo/cluster.rb:216]: Is this start_time used for something now?

File: lib/mongo/utils.rb [hunk 2]
```diff
@@ -87,5 +87,20 @@ def initialize(**opts)
...
+    # This function should be used if you need to measure time.
+    # @example Calculate elapsed time.
+    #   starting = Utils.monotonic_time
+    #   # do something time consuming
+    #   ending = Utils.monotonic_time
+    #   puts "It took #{(ending - starting).to_i} seconds"
+    #
+    # @see https://blog.dnsimple.com/2018/03/elapsed-time-with-ruby-the-right-way/
+    #
+    # @return [Float] seconds according to monotonic clock
```
Comment (p-mongo) [inline lib/mongo/utils.rb:100]: :+1:

File: lib/mongo/cluster.rb (hunk 1 above)
Comment (comandeo) [reply lib/mongo/cluster.rb:216]: It is used on line 244: `server.last_scan >= start_time`

The `last_scan` attribute of `Server` is public and explicitly documented with type `Time`. So, I am hesitant to change it.

Comment (p-mongo) [reply lib/mongo/cluster.rb:216]: I think there need to be two last_scan fields, the new one being monotonic_last_scan or last_scan_monotonic (or, given the naming conventions I prefer, I would suggest last_scan_monotime as a contraction from last_scan_monotonic_time).

Last scan is used in two places: 

- The server selector uses it for staleness calculations. The value there is compared to the server-reported time and must therefore remain to be a wallclock time.
-Cluster uses it for waiting in the Client constructor until all servers have been scanned. The time there should be monotonic.

The monotonic last scan can be `@api private`.

**Review by p-mongo — APPROVED** [review]

**Pull request merged**